extensions/amp-social-share/0.1/amp-social-share.js

```diff
@@ -0,0 +1,133 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {getSocialConfig} from './amp-social-share-config';
+import {addParamsToUrl} from '../../../src/url.js';
+import {elementByTag} from '../../../src/dom.js';
+
+/** @const {number} */
+const DEFAULT_WIDTH = 60;
+/** @const {number} */
+const DEFAULT_HEIGHT = 44;
+
+/** @const {string} */
+const CONTAINER_TAG = 'span';
+/** @const {string} */
+const LINK_TAG = 'a';
+/** @const {string} */
+const JSON_TAG = 'script';
+
+class AmpSocialShare extends AMP.BaseElement {
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  buildCallback() {
+
+    /** @private @const {Object} */
+    this.config_ = this.getElementConfig_(this.element);
+    AMP.assert(this.config_, 'The <script> config is required for ' +
+      '<amp-social-share> %s');
+
+    /** @private @const {!Element} */
+    this.type_ = AMP.assert(this.element.getAttribute('type'),
+        'The type attribute is required for <amp-social-share> %s',
+        this.element);
+    /** @private @const {!Object} */
+    this.typeConfig_ = getSocialConfig(this.type_);
+
+    /** @private @const {Element} */
+    this.container_ = this.renderShare_(this.config_);
+
+    /** @private @const {number} */
+    this.width = this.element.getAttribute('width') || DEFAULT_WIDTH;
+    /** @private @const {number} */
+    this.height = this.element.getAttribute('height') || DEFAULT_HEIGHT;
+
+    this.element.appendChild(this.container_);
+  }
+
+  /**
+   * Renders the share based on the provided config.
+   * @param {!Object} config
+   * @return {!Element}
+   */
+  renderShare_(config) {
+    const urlParams = {};
+
+    for (const param in this.typeConfig_['params']) {
+      const paramConf = this.typeConfig_['params'][param];
+      AMP.assert(!paramConf['required'] || param in config,
+        param + ' is a required configuration item for ' + this.type_);
+
+      if (param in config) {
+        const value = config[param];
+        if ('maxlength' in paramConf) {
+          const maxlength = paramConf['maxlength'];
+          AMP.assert(value.length < maxlength,
+            param + ' cannot exceed ' + maxlength);
+        }
+        urlParams[paramConf['param']] = value;
+      } else if (paramConf['type'] == 'fixed') {
+        urlParams[paramConf['param']] = paramConf['value'];
+      }
+    }
+    let container = elementByTag(this.element, CONTAINER_TAG);
```

Comment:
Could you please explain who these correlate with examples above? I have a bit hardtime following it. It seems like you search in the whole subtree to find a `span` or an `a`. Is that right?